rtos/mbed_rtos_storage.h

```diff
@@ -0,0 +1,51 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2017 ARM Limited
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#ifndef MBED_RTOS_STORAGE_H
+#define MBED_RTOS_STORAGE_H
+
+/** \addtogroup rtos */
+/** @{*/
+
+/** @brief RTOS primitives storage types for RTX
+
+ Types defined in this file should be utilized, when the direct RTOS C API usage is required, to provide backing memory
+ for internal RTX data. Allocated object should be wrapped in attribute struct and passed to os*New call, for details
+ see CMSIS-RTOS2 documentation.
+
+ @note
+ This file breaks abstraction layers and uses RTX internal types, but it limits the contamination to single, RTOS
+ implementation specific, header file, therefore limiting scope of possible changes.
+ */
+
+#include "rtx_lib.h"
+
+typedef os_mutex_t mbed_rtos_storage_mutex_t;
```

Comment:
I can't find much info for this type ``os_mutex_t`` but that looks like already generic type (rtx redefines it to their own RTX related thing). I tried to look at cmsis5 docs, it is not there, neither in cmsis5 for freertos, could not find it there so this really might be rtx specific storage. Wondering why they do it `` #define os_mutex_t          osRtxMutex_t`` 😖 How does these typedef solve our problem? I assume what is missing in cmsis is ``osMutexStorage_t``  type?

Taken from https://github.com/ARM-software/CMSIS-FreeRTOS/blob/74b3686c5b6b1816e1f4cc53e9d883f96e3d4cd9/CMSIS/RTOS2/FreeRTOS/Include1/cmsis_os.h#L297 :

```
/// Mutex Definition structure contains setup information for a mutex.
#if (osCMSIS < 0x20000U)
typedef struct os_mutex_def {
  uint32_t                     dummy;   ///< dummy value
} osMutexDef_t;
#else
#define osMutexDef_t osMutexAttr_t
#endif
```

why they just don't do one composite data that would hide all these (id, attribute and storage). Not like pass me attribute, I return you id and hey allocate storage if you need to? This is really confusing, might be missing something here 😲